### PR TITLE
discovery: add support for Managed Identity authentication in Azure SD

### PR DESCRIPTION
**Pull request by marratj**
Azure supports Managed Identity authentication for a while now: https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview

This PR is to add support in the Azure SD for Managed Identity auth as an alternative to OAuth with an Azure AD Service Principal.

### CODE REVIEW

File: discovery/azure/azure.go [hunk 1]
```diff
@@ -170,13 +172,32 @@ func createAzureClient(cfg SDConfig) (azureClient, error) {
 	resourceManagerEndpoint := env.ResourceManagerEndpoint
 
 	var c azureClient
+
 	oauthConfig, err := adal.NewOAuthConfig(activeDirectoryEndpoint, cfg.TenantID)
```
Comment (simonpasquier) [inline discovery/azure/azure.go]: shouldn't it be moved to the "OAuth" switch case too?

Comment (marratj) [reply discovery/azure/azure.go]: You're right, just changed it.

File: discovery/azure/azure.go [hunk 2]
```diff
@@ -170,13 +172,32 @@ func createAzureClient(cfg SDConfig) (azureClient, error) {
...
+		if err != nil {
+			return azureClient{}, err
+		}
+
+		spt, err = adal.NewServicePrincipalToken(*oauthConfig, cfg.ClientID, string(cfg.ClientSecret), resourceManagerEndpoint)
+		if err != nil {
+			return azureClient{}, err
+		}
+	default:
+		return azureClient{}, fmt.Errorf("Unsupported authentication_type set. Needs to be 'OAuth' or 'ManagedServiceIdentity'")
```
Comment (simonpasquier) [inline discovery/azure/azure.go]: This should be handled by `UnmarshalYAML()` at L86.

Comment (marratj) [reply discovery/azure/azure.go]: Right, as the SubscriptionID check is also handled there, understood.

I have put the check into UnmarshalYAML() and removed the default case.

Hope this looks ok now :)

File: discovery/azure/azure.go [hunk 3]
```diff
@@ -91,6 +93,11 @@ func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	if c.SubscriptionID == "" {
 		return fmt.Errorf("Azure SD configuration requires a subscription_id")
 	}
+
+	if c.AuthenticationMethod != "OAuth" && c.AuthenticationMethod != "ManagedServiceIdentity" {
+		return fmt.Errorf("Unsupported authentication_type set. Needs to be 'OAuth' or 'ManagedServiceIdentity'")
```
[the remark this answers is not in the document]
Comment (marratj) [reply discovery/azure/azure.go]: Ok, I have changed it so it displays the invalid value and shows the available choices.

However, on writing a test on the check I'm honestly a bit lost. I actually don't have any real experience in writing unit tests (this is one of my first PRs). I tried to look at how other SD plugins test for this stuff, but I see that many don't even have tests (like ec2, dns & gce).

For any hints on how to structure that test I would be grateful :-)

File: docs/configuration/configuration.md [hunk 4]
```diff
@@ -274,13 +274,16 @@ See below for the configuration options for Azure discovery:
...
 # The Azure environment.
 [ environment: <string> | default = AzurePublicCloud ]
-# The subscription ID.
+# The authentication method
+# Either OAuth or ManagedServiceIdentity (https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview)
+[ authentication_method: <string> | default = OAuth]
+# The subscription ID. Always required.
 subscription_id: <string>
-# The tenant ID.
+# The tenant ID. Only required with authentication_method OAuth.
```
[the remark this answers is not in the document]
Comment (marratj) [reply docs/configuration/configuration.md]: It's a good idea, but I agree that we might want to have it as a separate PR.

File: discovery/azure/azure.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (marratj) [reply discovery/azure/azure.go]: OK, so I added also a bad config test similar to the OpenStack SD one.

**Comment by simonpasquier** [comment]
:+1:

**Comment by simonpasquier** [comment]
@brian-brazil can you have a look at this?

**Comment by marratj** [comment]
Any chance it can be merged in time for 2.5? Or is anything missing? :-)

File: discovery/azure/azure_test.go [hunk 5]
```diff
@@ -20,6 +20,12 @@ import (
 	"github.com/Azure/azure-sdk-for-go/arm/compute"
 )
 
+func TestAuthenticationMethodIsValid(t *testing.T) {
```
Comment (simonpasquier) [inline discovery/azure/azure_test.go]: This test doesn't bring much, I'd remove it.

File: docs/configuration/configuration.md [hunk 6]
```diff
@@ -274,13 +274,16 @@ See below for the configuration options for Azure discovery:
 # The information to access the Azure API.
 # The Azure environment.
 [ environment: <string> | default = AzurePublicCloud ]
-# The subscription ID.
+# The authentication method
```
Comment (simonpasquier) [inline docs/configuration/configuration.md]: I'd add a blank line as the following parameters relate to authentication only.

File: discovery/azure/azure.go [hunk 7]
```diff
@@ -63,21 +63,23 @@ var (
 
 	// DefaultSDConfig is the default Azure SD configuration.
 	DefaultSDConfig = SDConfig{
-		Port:            80,
-		RefreshInterval: model.Duration(5 * time.Minute),
-		Environment:     azure.PublicCloud.Name,
+		Port:                 80,
+		RefreshInterval:      model.Duration(5 * time.Minute),
+		Environment:          azure.PublicCloud.Name,
+		AuthenticationMethod: "OAuth",
```
[the remark this answers is not in the document]
Comment (marratj) [reply discovery/azure/azure.go]: I have defined as constants (I hope the place I put them in is ok).

Also, I have changed the name of the Auth Method from ManagedServiceIdentity to ManagedIdentity to reflect the fact that Microsoft has renamed it.

**Comment by marratj** [comment]
@simonpasquier @brian-brazil I think I have addressed all your comments, can you please have another look? :)

Also, I have changed the name of the Auth Method from ManagedServiceIdentity to ManagedIdentity to reflect the fact that Microsoft has renamed it.

**Comment by marratj** [comment]
rebased again after #4907 

@simonpasquier any chance we get this PR merged soon, as it is already 3 months old and was approved twice by you already?

**Comment by marratj** [comment]
@brian-brazil any objections against merging this PR?

File: docs/configuration/configuration.md [hunk 8]
```diff
@@ -275,13 +275,17 @@ See below for the configuration options for Azure discovery:
...
-# The subscription ID.
+
+# The authentication method, either OAuth or ManagedIdentity.
+# See https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview
+[ authentication_method: <string> | default = OAuth]
+# The subscription ID. Always required.
 subscription_id: <string>
-# The tenant ID.
+# The tenant ID. Only required with authentication_method OAuth.
 tenant_id: <string>
```
Comment (brian-brazil) [inline docs/configuration/configuration.md]: This should be marked optional. And below.

Comment (marratj) [reply docs/configuration/configuration.md]: They are kinda semi-optional, as they are only optional when using ManagedIdentity (in which case they are not even used).

However, they are required when using OAuth as authentication method (which is the default) and Prometheus will fail hard when they are not set in this case (introduced with #4907 )

Comment (brian-brazil) [reply docs/configuration/configuration.md]: They're only required if they're always required.

Comment (marratj) [reply docs/configuration/configuration.md]: OK, I have marked them now as optional in the docs.

**Comment by marratj** [comment]
@brian-brazil done

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**